liberum_core/src/swarm_runner/behaviour/object_sender.rs

```diff
@@ -0,0 +1,228 @@
+            if let ObjectEnum::Result(r) = parser::parse_typed(response.object).await.unwrap() {
+                let _ = sender.send(Ok(r));
+            } else {
+                let _ = sender.send(Err(anyhow!("Failed to parse result object")));
+            }
+        }
+    }
+
+    fn respond_err(
+        &mut self,
+        request: ObjectSendRequest,
+        response_channel: ResponseChannel<ObjectResponse>,
+    ) {
+        let _ = self.swarm.behaviour_mut().object_sender.send_response(
+            response_channel,
+            ObjectResponse {
+                object: proto::ResultObject { result: Err(()) }.into(),
+                object_id: request.object_id,
+            },
+        );
+    }
+
+    async fn handle_request_plain_file(
+        &mut self,
+        obj: PlainFileObject,
+        id: proto::Hash,
+        request: ObjectSendRequest,
+        _request_id: InboundRequestId,
+        response_channel: ResponseChannel<ObjectResponse>,
+    ) {
+        self.put_object_into_vault(obj.into()).await.unwrap();
+        let qid = self
+            .swarm
+            .behaviour_mut()
+            .kademlia
+            .start_providing(kad::RecordKey::from(id.bytes.to_vec()))
+            .unwrap();
+        self.behaviour
+            .pending_object_start_providing
+            .insert(qid, (request.object_id, response_channel));
+    }
+
+    async fn handle_request_query(
+        &mut self,
+        query: QueryObject,
+        _id: proto::Hash,
+        request: ObjectSendRequest,
+        _request_id: InboundRequestId,
+        response_channel: ResponseChannel<ObjectResponse>,
+    ) {
+        let query = parser::parse_typed(query.query_object).await;
+        if let Err(e) = query {
+            debug!(err = format!("{e}"), "query_object couldnt get parsed");
+            self.respond_err(request, response_channel);
+            return;
+        }
+        let query = query.unwrap();
+
+        let query = match query {
+            parser::ObjectEnum::SimpleIDQuery(query) => parser::ObjectEnum::SimpleIDQuery(query),
+            _ => {
+                error!("query_object TypedObject is not SimpleIDQuery {query}");
+                return;
+            }
+        };
+
+        match query {
+            parser::ObjectEnum::SimpleIDQuery(query) => {
+                let obj = self.get_object_from_vault(query.id.clone());
+                if let None = obj {
+                    self.respond_err(request, response_channel);
+                    return;
+                }
+                let obj = obj.unwrap();
+                let id = proto::Hash::try_from(&obj).unwrap();
+                if query.id != id {
+                    error!(
+                        received_id = bs58::encode(&query.id.bytes).into_string(),
+                        computed_id = bs58::encode(&id.bytes).into_string(),
+                        "ids dont match"
```

Comment:
Be more specific here.

Reply:
https://github.com/Liberum-Neto/liberum-neto/pull/56/commits/0a49bba4233068604cc2edf5ff6c123e4a292dfa